aptos/token_bridge/sources/bridge_state.move

```diff
@@ -118,6 +147,92 @@ module token_bridge::bridge_state {
+        assert!(is_wrapped_asset(token), E_IS_NOT_WRAPPED_ASSET);
+        assert!(exists<CoinCapabilities<CoinType>>(@token_bridge), E_COIN_CAP_DOES_NOT_EXIST);
+        let caps = borrow_global<CoinCapabilities<CoinType>>(@token_bridge);
+        let mint_cap = &caps.mint_cap;
+        let coins = mint<CoinType>(amount, mint_cap);
+        coins
+    }
+
+    // this function is called in tandem with bridge_implementation::create_wrapped_coin_type
+    public entry fun create_wrapped_coin<CoinType>(vaa: vector<u8>) acquires State{
+        //TODO: parse and verify and replay protect
+        //let vaa = parse_verify_and_replay_protect(vaa);
+        let vaa = parse_and_verify(vaa);
+        let _asset_meta: AssetMeta = bridge_structs::parse_asset_meta(vaa::get_payload(&vaa));
+
+        // fetch signer_cap and create signer for CoinType
+        let token_address = account_address(&type_of<CoinType>());
+        let wrapped_coin_signer_caps = &borrow_global<State>(@token_bridge).wrapped_asset_signer_capabilities;
+        let coin_signer_cap = table::borrow(wrapped_coin_signer_caps, to_bytes(&token_address));
+        let coin_signer = create_signer_with_capability(coin_signer_cap);
+
+        // initialize new coin using CoinType
+        let name = bridge_structs::get_name(&_asset_meta);
+        let symbol = bridge_structs::get_symbol(&_asset_meta);
+        let decimals = bridge_structs::get_decimals(&_asset_meta);
+        let monitor_supply = false;
+        let (burn_cap, freeze_cap, mint_cap) = initialize<CoinType>(&coin_signer, utf8(name), utf8(symbol), decimals, monitor_supply);
+
+        // update the following two mappings in State
+        // 1. (native chain, native address) => wrapped address
+        // 2. wrapped address => (native chain, native address)
+        let token_address = bridge_structs::get_token_address(& _asset_meta);
+        let token_chain = bridge_structs::get_token_chain(& _asset_meta);
+        let native_info = NativeInfo {token_address: token_address, token_chain: token_chain};
+        set_native_info(token_address, &native_info);
+        set_wrapped_asset(&native_info, token_address);
+
+        // store coin capabilities
+        let _token_bridge_signer = token_bridge_signer();
+        let coin_caps = CoinCapabilities<CoinType> { mint_cap: mint_cap, freeze_cap: freeze_cap, burn_cap: burn_cap};
+        move_to(&_token_bridge_signer, coin_caps);
+
+        vaa::destroy(vaa);
+    }
+
+    // transfer a native or wraped token from sender to token_bridge
+    public fun transfer_tokens<CoinType>(_sender: &signer, _amount: u128, _arbiter_fee: u128): TransferResult acquires State {
+        // TODO: assertions and checks,
+        // check if CoinType is registered with @token_bridge
+        let token_address = hash_type_info<CoinType>();
+
+        // transfer tokens from sender to token_bridge
+        transfer<CoinType>(_sender, @token_bridge, (_amount as u64));
```

Comment:
this might fail if the token bridge is not registered for the coin yet

Reply:
instead of casting `amount as u64` the function should just take a u64 amount as an argument -- precise type information should always be pushed outwards as much as possible

Reply:
left the check for coin registration as a todo, but will do the assertion today!